src/algebraic_geometry/properties.lean

```diff
@@ -83,6 +83,97 @@ begin
   rw [comp_apply, e', map_zero]
 end
 
+lemma is_reduced_of_open_immersion {X Y : Scheme} (f : X ⟶ Y) [H : is_open_immersion f]
+  [is_reduced Y] : is_reduced X :=
+begin
+  constructor,
+  intro U,
+  have : U = (opens.map f.1.base).obj (H.base_open.is_open_map.functor.obj U),
+  { ext1, exact (set.preimage_image_eq _ H.base_open.inj).symm },
+  rw this,
+  exact is_reduced_of_injective (inv $ f.1.c.app (op $ H.base_open.is_open_map.functor.obj U))
+    (as_iso $ f.1.c.app (op $ H.base_open.is_open_map.functor.obj U) : Y.presheaf.obj _ ≅ _).symm
+      .CommRing_iso_to_ring_equiv.injective
+end
+
+local attribute [elementwise] category_theory.is_iso.hom_inv_id
+
+lemma basic_open_eq_of_affine {R : CommRing}
+  (f : R) :
+  RingedSpace.basic_open (Spec.to_SheafedSpace.obj (op R)) ((Spec_Γ_identity.app R).inv f) =
+    prime_spectrum.basic_open f :=
+begin
+  ext,
+  change ↑(⟨x, trivial⟩ : (⊤ : opens _)) ∈
+    RingedSpace.basic_open (Spec.to_SheafedSpace.obj (op R)) _ ↔ _,
+  rw RingedSpace.mem_basic_open,
+  suffices : is_unit (structure_sheaf.to_stalk R x f) ↔ f ∉ prime_spectrum.as_ideal x,
+  { exact this },
+  erw [← is_unit_map_iff (structure_sheaf.stalk_to_fiber_ring_hom R x),
+    structure_sheaf.stalk_to_fiber_ring_hom_to_stalk],
+  exact (is_localization.at_prime.is_unit_to_map_iff
+    (localization.at_prime (prime_spectrum.as_ideal x)) (prime_spectrum.as_ideal x) f : _)
+end
+
+lemma basic_open_eq_of_affine' {R : CommRing}
+  (f : (Spec.to_SheafedSpace.obj (op R)).presheaf.obj (op ⊤)) :
+  RingedSpace.basic_open (Spec.to_SheafedSpace.obj (op R)) f =
+    prime_spectrum.basic_open ((Spec_Γ_identity.app R).hom f) :=
+begin
+  convert basic_open_eq_of_affine ((Spec_Γ_identity.app R).hom f),
+  exact (coe_hom_inv_id _ _).symm
+end
+
+lemma eq_zero_of_basic_open_empty {X : Scheme} [is_reduced X] {U : opens X.carrier}
+  (s : X.presheaf.obj (op U)) (hs : X.to_LocallyRingedSpace.to_RingedSpace.basic_open s = ∅) :
+  s = 0 :=
+begin
```

Comment:
I feel like this should have a shorter proof. Most of the proof is happening in about 10 lines, the rest seems to be moving through a "generic strategy" to reduce to a local question. Can we formulate an induction principle that abstracts this approach?

Reply:
The proof is now refactored. It looks better and more straight forward, though I'm afraid it didn't really get any shorter.